openadapt/db/crud.py

```diff
@@ -420,71 +457,75 @@ def save_screenshot_diff(screenshots: list[Screenshot]) -> list[Screenshot]:
 
 
 def get_screenshots(
-    recording: Recording, session: sa.orm.Session = None
+    db: SaSession,
+    recording: Recording,
 ) -> list[Screenshot]:
     """Get screenshots for a given recording.
 
     Args:
+        db (sa.orm.Session): The database session.
         recording (Recording): The recording object.
-        session (sa.orm.Session, optional): The database session. Defaults to None.
 
     Returns:
         list[Screenshot]: A list of screenshots for the recording.
     """
-    screenshots = _get(Screenshot, recording.timestamp, session)
+    screenshots = _get(db, Screenshot, recording.timestamp)
 
     for prev, cur in zip(screenshots, screenshots[1:]):
         cur.prev = prev
     if screenshots:
         screenshots[0].prev = screenshots[0]
 
     if config.SAVE_SCREENSHOT_DIFF:
-        screenshots = save_screenshot_diff(screenshots)
+        screenshots = save_screenshot_diff(db, screenshots)
     return screenshots
 
 
 def get_window_events(
-    recording: Recording, session: sa.orm.Session = None
+    db: SaSession,
+    recording: Recording,
 ) -> list[WindowEvent]:
     """Get window events for a given recording.
 
     Args:
+
         recording (Recording): The recording object.
-        session (sa.orm.Session, optional): The database session. Defaults to None.
 
     Returns:
         list[WindowEvent]: A list of window events for the recording.
     """
-    return _get(WindowEvent, recording.timestamp, session)
+    return _get(db, WindowEvent, recording.timestamp)
 
 
-def new_session() -> None:
-    """Create a new database session.
-
-    This was necessary because the database session was not being closed
-    properly, and the database would become locked.
-    """
-    global db
-    if db:
-        db.close()
-    db = Session()
-
-
-def get_new_session() -> sa.orm.Session:
+def get_new_session(
+    read_only: bool = False,
+    read_and_write: bool = False,
+) -> sa.orm.Session:
     """Get a new database session.
 
+    Args:
+        read_only (bool): Whether to open the session in read-only mode.
+        read_and_write (bool): Whether to open the session in read-and-write mode.
+
     Returns:
         sa.orm.Session: A new database session.
     """
+    assert not (
+        read_only and read_and_write
+    ), "Cannot be both read-only and read-and-write."
+    assert read_only or read_and_write, "Must be either read-only or read-and-write."
+    if read_only:
+        return ReadOnlySession()
     return Session()
```

Comment:
I wonder if there is a simpler approach.

Can you please evaluate each of these options suggested by ChatGPT to see if they resolve the issue without using a separate session?

---

1. **Connection Configuration**:
   - Use `isolation_level` to set the transaction behavior in SQLAlchemy. For SQLite, setting the `isolation_level` to `"IMMEDIATE"` can be useful. It starts a transaction immediately but doesn’t lock the database until a write operation is performed. This allows reads to happen concurrently until a write is needed.
   ```python
   engine = create_engine('sqlite:///yourdb.db', isolation_level="IMMEDIATE")
   NonLockingSession = sessionmaker(bind=engine, autoflush=False, autocommit=False)
   ```

2. **Database File Configuration**:
   - Ensure that your SQLite database file is on a fast storage medium to reduce the time any lock is held.
   - Consider using the `WAL` (Write-Ahead Logging) mode if your application can handle it. WAL can allow concurrent reads and writes. Enable WAL mode using a command like this after establishing a connection:
   ```python
   engine.execute("PRAGMA journal_mode=WAL;")
   ```